### ISSUE TITLE
Code breaks with v2.0.1

### ISSUE THREAD

**Issue by hauptmech** [warmcat/libwebsockets#542]
With the v2.0.1 update the following code fails with:
```
lwsts[14437]: ****** 980008c0 Sending new, pending truncated ...
```

```c
// in static int callback_http() ....
	switch (reason) {
	case LWS_CALLBACK_HTTP:
<snip>
// Send state_information.js as a special file
		if (!strcmp((const char *)in, "/state_information.js")) {
			p = buffer + LWS_SEND_BUFFER_PRE_PADDING;
			end = p + sizeof(buffer) - LWS_SEND_BUFFER_PRE_PADDING;

			//Prepare header
			if (lws_add_http_header_status( wsi, 200, &p, end))
				return 1;
			if (lws_add_http_header_by_token( wsi,WSI_TOKEN_HTTP_SERVER,
				    	(unsigned char *)"libwebsockets", 13, &p, end))
				return 1;
			if (lws_add_http_header_by_token( wsi,WSI_TOKEN_HTTP_CONTENT_TYPE,
				    	(unsigned char *)"text/javascript", 15, &p, end))
				return 1;
			if (lws_add_http_header_content_length( wsi, state_information.size(), &p, end))
				return 1;
			if (lws_finalize_http_header( wsi, &p, end))
				return 1;

			
			n = lws_write(wsi,
					buffer + LWS_SEND_BUFFER_PRE_PADDING,
					p - (buffer + LWS_SEND_BUFFER_PRE_PADDING),
					LWS_WRITE_HTTP_HEADERS);
			
			n = lws_write(wsi,
					(unsigned char*)state_information.c_str(),
					state_information.size(),
					LWS_WRITE_HTTP_FINAL);
			
			if (n < 0 || ((n > 0) && lws_http_transaction_completed(wsi)))
				return -1;
			break;
		}
```
The code seemed to work fine in v2.0.0.
I have no idea whether a bug was introduced in v2.0.1 or if my code was written wrong. 
I'm reverting to v2.0.0. Sorry I don't have time to track down more info on this....

**Comment by lws-team** [comment]
There is no context for the code you pasted.  But my guess is this has nothing to do with any change on lws side.

Generally, any lws_write() may find that it has more to send than the kernel is willing to take.  In that case, he will privately take a malloc'd copy of the remainder and take care of completing the sending of that before letting the user application know that it can accept any more.

Whenever you see two lws_write()s back-to-back it's possible the first one was not able to complete and enetered this moe where it will complete the remaining write privately.  In that case, the second write is illegal.

Unless you have some external knowledge that the second write will be accepted, you should ask for a writable callback after the first write, and do the second write then.

But from what you pasted, it doesn't even explain at what point the first write is done.

**Comment by hauptmech** [comment]
Thank you, that's helpful (needing a writable_callback in between lws_write()s). The snippet is in the http callback with reason LWS_CALLBACK_HTTP. The code runs fine in v2.0.0 so it's definately on the lws side; if I was abusing the api and getting lucky, it was very consistent luck.  I'll revert for now and revisit this when I can. You can close this if you don't think something bad was introduced in the lws update.

**Comment by lws-team** [comment]
> The code runs fine in v2.0.0 so it's definately on the lws side

Wow... 

> revisit this when I can

Yeah... don't strain yourself.

> You can close this if you don't think something bad was introduced in the lws update.

Thanks!

**Comment by hauptmech** [comment]
Sorry, I did a really poor job of communicating there.

First, I really appreciate your quick response even when I gave so little context. 

I also really appreciate lws being here. It's allowed me to do things that would normally be beyond what I can do. 

My code didn't change, lws code did change, stuff broke. I posted the issue in case it was helpful for you guys to know that. I was not looking for you to fix my code. I was not trying to desparage lws. 

>  I'll revert for now and revisit this when I can. 

"this" being making my code use the lws api correctly. 

And I should have ended with something more like "It sounds like this issue should be closed, are you sure it's not a lws problem?".

Thanks again for lws and your help.
